In [1]:
from sklearn import datasets, metrics
from sklearn.model_selection import train_test_split, KFold, GridSearchCV
from sklearn.ensemble import GradientBoostingRegressor
import warnings
warnings.filterwarnings("ignore")

wine = datasets.load_wine()

x_train, x_test, y_train, y_test = train_test_split(wine.data, wine.target, test_size=0.25, random_state=42)

In [2]:
# MSE before hyper-parameter tunning
clf = GradientBoostingRegressor(random_state=7)
clf.fit(x_train, y_train)
y_pred = clf.predict(x_test)
print(metrics.mean_squared_error(y_test, y_pred))

0.14672894294872418


In [3]:
# set the hyper-parameter that is going to train
n_estimators = [100, 200, 300]
max_depth = [1, 3, 5]
param_grid = dict(n_estimators=n_estimators, max_depth=max_depth)

## search the best hyper-parameter (n_jobs=-1 use all cpus to do parallel computing)
grid_search = GridSearchCV(clf, param_grid, scoring="neg_mean_squared_error", n_jobs=-1, verbose=1)
grid_result = grid_search.fit(x_train, y_train)

print("Best Accuracy: %f using %s" % (grid_result.best_score_, grid_result.best_params_))

Fitting 3 folds for each of 9 candidates, totalling 27 fits


[Parallel(n_jobs=-1)]: Using backend LokyBackend with 8 concurrent workers.


Best Accuracy: -0.081166 using {'max_depth': 1, 'n_estimators': 100}


[Parallel(n_jobs=-1)]: Done  27 out of  27 | elapsed:    3.2s finished


In [4]:
# MSE after hyper-parameter tunning
clf_bestparam = GradientBoostingRegressor(max_depth=grid_result.best_params_['max_depth'],
                                           n_estimators=grid_result.best_params_['n_estimators'])

clf_bestparam.fit(x_train, y_train)
y_pred = clf_bestparam.predict(x_test)
print(metrics.mean_squared_error(y_test, y_pred))

0.054537686000142374


The best hyper-parameters tunned are 'max_depth' = 1 and 'n_estimators' = 100, which decrease the MSE from 0.1467 to 0.0545.